Fixed #35444 -- Added generic support for Aggregate.order_by

camuthig opened the pull request:
# Trac ticket number

ticket-35444

# Branch description
There are three commits to this draft that represent the three phases of the change:

1. Deprecate the `ordering` argument on the PostgreSQL aggregate functions and replace it with `order_by` to create a consistent naming convention
2. Add generic support for `Aggregate.order_by` and migrate the Postgres-specific `StringAgg` class to the shared `aggregates` module, deprecating the Postgres version.
3. Deprecate the `OrderableAggMixin`

# Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.

camuthig commented:
@charettes I've made some progress on the suggestions. I have the three phases broken out into unique commits and have added some of the documentation and copied the tests from the postgres module into a new class that should run against all of the database backends.

I ran into two issues in the process that I wanted to get your opinion on.

1. Python is compiled against Sqlite3 version 3.37.0, which doesn't have support for the order by yet. So the ordering tests aren't running for that. I think I would need to compile Python from source and point it to a specific version of Sqlite3 to get past this, but I'm not sure if you have any thoughts.
2. Sqlite3 doesn't behave well when you combine `distinct` and a delimiter value. It throws an error of `SQLite doesn't support DISTINCT on aggregate functions accepting multiple arguments.` So it treats the delimiter as a second argument. In Sqlite if you don't provide a delimiter it will use a comma as a default. I was trying to build it such that if the user sets the delimiter to a comma and is using Sqlite as a backend, then the delimiter is ignored. However, I could figure out how to make that work. I tried a few things around custom SQL, using a class attribute instead of making it part of `expressions`, etc. but I couldn't quite get it right. If you can see a good way to treat that, I think it would be awesome to make distinct work in Sqlite for at least some scenarios, if not all.

As a side note, I named it `StringAgg` because that is what I am used to, and it feels like the more descriptive function name between that a `GROUP_CONCAT`, but I'm open to changing that to whatever you think is best. I guess would could always have an alias class `class GroupConcat(StringAgg): pass` if we really wanted to make it smooth for developers from either environment, but that doesn't seem to pass the smell test on "one way to do something"

charettes commented:
Hello @camuthig, thanks for spinning this up!

> Python is compiled against Sqlite3 version 3.37.0, which doesn't have support for the order by yet. So the ordering tests aren't running for that. I think I would need to compile Python from source and point it to a specific version of Sqlite3 to get past this, but I'm not sure if you have any thoughts.

It's possible CI might not be be setup here but if you're a *nix setup [you can use `LD_PRELOAD` to point at any SQLite version](https://til.simonwillison.net/sqlite/ld-preload#user-content-building-the-library). You don't have to build from source as [the SQLite project provides pre-built binaries](https://sqlite.org/download.html).

> Sqlite3 doesn't behave well when you combine distinct and a delimiter value. It throws an error of SQLite doesn't support DISTINCT on aggregate functions accepting multiple arguments. So it treats the delimiter as a second argument. In Sqlite if you don't provide a delimiter it will use a comma as a default. I was trying to build it such that if the user sets the delimiter to a comma and is using Sqlite as a backend, then the delimiter is ignored. However, I could figure out how to make that work. I tried a few things around custom SQL, using a class attribute instead of making it part of expressions, etc. but I couldn't quite get it right. If you can see a good way to treat that, I think it would be awesome to make distinct work in Sqlite for at least some scenarios, if not all.

Well [that's odd for sure](https://www.sqlite.org/lang_aggfunc.html#group_concat).

> In any aggregate function that takes a single argument, that argument can be preceded by the keyword DISTINCT. In such cases, duplicate elements are filtered before being passed into the aggregate function. For example, the function "count(distinct X)" will return the number of distinct values of column X instead of the total number of non-null values in column X. 

Usually when we run into these edge cases we [add a database feature](https://github.com/django/django/blob/dfd63ff43408e7901cc214b0482a7f844244d439/django/db/backends/base/features.py#L5) (e.g. `supports_aggregate_distinct_multiple_argument`) and use for two purpose.

1. Adjust `Aggregate.as_sql` to raise an error when `self.distinct and not connection.features.supports_aggregate_distinct_multiple_argument and len(super().get_expressions()) > 1`
2. Add tests for `@skipUnlessDBFeature("supports_aggregate_distinct_multiple_argument")` that cover the backends that support it and `@skipIfDBFeature` that makes sure the proper exception is raised.

> As a side note, I named it StringAgg because that is what I am used to, and it feels like the more descriptive function name between that a GROUP_CONCAT, but I'm open to changing that to whatever you think is best. 

I think `StringAgg` is fine as that's the name most backends use (Postgres, SQLite, SQLServer)

charettes commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -49,17 +51,15 @@ class BoolOr(Aggregate):
+        warnings.warn(
+            "The PostgreSQL specific StringAgg function is deprecated. Use "
+            "django.db.models.aggregate.StringAgg instead.",
+            category=RemovedInDjango60Warning,
```
We might have to adjust the `stacklevel` here so the warning points at the location of `StringAgg` instantiation. I think `stacklevel=1` is the right one?

charettes commented on django/contrib/postgres/aggregates/mixins.py:
```diff
@@ -1,47 +1,24 @@
+        if ordering:
+            warnings.warn(
+                "The ordering argument is deprecated. Use order_by instead.",
+                category=RemovedInDjango60Warning,
```
I think this will need `stacklevel` as well.

charettes commented on django/contrib/postgres/aggregates/mixins.py:
```diff
@@ -1,47 +1,24 @@
-        for source_expr in source_exprs:
-            source_params += compiler.compile(source_expr)[1]
+    def __init__(self, *expressions, ordering=(), **extra):
+        warnings.warn(
+            "The OrderableAggMixin is deprecated. Use Aggregate and allow_order_by "
+            "instead.",
+            category=RemovedInDjango60Warning,
+        )
```
I wonder if we should move this [to `__init_subclass__` instead](https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__) and do something like

```python
class _DeprecatedOrdering:
    def __init__(self, *expressions, ordering=(), **extra):
        if ordering:
            warnings.warn(
                "The ordering argument is deprecated. Use order_by instead.",
                category=RemovedInDjango60Warning,
            )
        super().__init__(*expressions, **extra)

class OrderableAggMixin(_DeprecatedOrdering):
    allow_order_by = True

    def __init_subclass__(cls, /, *args, **kwargs):
        warnings.warn(
            "OrderableAggMixin is deprecated. Use Aggregate and allow_order_by "
            "instead.",
            category=RemovedInDjango60Warning,
        )
        super().__init_subclass__(*args, **kwargs)
```

And then in `postgres.aggregates` subclass `_DeprecatedOrdering` instead and explicitly set `allow_order_by = True` as well.

That would make any subclassing of `OrderableAggMixin` emit the deprecation warning while allowing `postgres.aggregates` to benefit from the deprecation period.

charettes commented on django/db/models/aggregates.py:
```diff
@@ -196,6 +245,52 @@ def _get_repr_options(self):
+    output_field = TextField()
+
+    def __init__(self, expression, delimiter, **extra):
+        delimiter = Value(str(delimiter))
```
We'll want to force the delimiter to be specified as a `Value`. This current form disallows using another expression as the delimiter such as a column reference.

charettes commented on django/db/models/aggregates.py:
```diff
@@ -196,6 +245,52 @@ def _get_repr_options(self):
+        delimiter = Value(str(delimiter))
+        super().__init__(expression, delimiter, **extra)
+
+    def as_sql(self, compiler, connection, **extra_context):
+        return super().as_sql(compiler, connection, **extra_context)
```
I think this one can go?

charettes commented on django/db/models/aggregates.py:
```diff
@@ -196,6 +245,52 @@ def _get_repr_options(self):
+        return self.as_sql(
+            compiler,
+            connection,
+            function="GROUP_CONCAT",
```
I _think_ [we can use `STRING_AGG` on SQLite as well](https://www.sqlite.org/lang_aggfunc.html#group_concat)?

camuthig replied:
I expected that as well, but I was getting errors that the function wasn't defined. It turns out, `STRING_AGG` was added in 3.44.0 at the same time as the `order_by` was added to the behaviors. 

https://www.sqlite.org/changes.html#version_3_44_0

Since my local sqlite versions have been 3.37, it was failing. Not sure if we have a preference on which to use for sqlite in this case. `GROUP_CONCAT` feels like it is a bit more flexible since it is older, but I don't know if the behaviors are exactly the same or not under the hood. I'll have to figure that out once I get a newer version of sqlite hooked up to my tests.

charettes replied:
I missed that, let's stick to `GROUP_CONCAT` then. I guess we could add a `connection.get_database_version() < (3, 44)` check so eventually when we drop support for < 3.44 we'll drop that code at the same time.

charettes commented on tests/aggregation/tests.py:
```diff
@@ -2158,6 +2177,18 @@ def test_having_with_no_group_by(self):
 
+    def test_string_agg(self):
+        values = Publisher.objects.aggregate(
+            stringagg=StringAgg("name", delimiter=";")
```
```suggestion
            stringagg=StringAgg("name", delimiter=Value(";"))
```

It'd be great to add testing for expression delimiters as well!

charettes commented on tests/aggregation/tests.py:
```diff
@@ -2359,3 +2390,256 @@ def test_aggregate_combined_queries(self):
         self.assertEqual(qs.count(), 3)
+
+
+class StringAggTestModel(Model):
+    char_field = CharField(max_length=30, blank=True)
+    text_field = TextField(blank=True)
+    json_field = JSONField(null=True)
+
+
+class StatTestModel(Model):
+    int1 = IntegerField()
+    int2 = IntegerField()
+    related_field = ForeignKey(StringAggTestModel, SET_NULL, null=True)
```
Try reuse the existing models defined in `aggregation/models.py` instead. We can't add new tables/models for each new features otherwise the suite slowly gets slower.

camuthig replied:
These tests are taken directly from the PostgreSQL specific of these tests. My thoughts were

1. We are deprecating that class in 6.0, so we will delete all of the StringAgg tests from that module at that time
2. Those tests exist for a reason, even if I am not familiar with them, so I didn't want to drop any and cause a regression

This is why I have some tests that may seem very specific, like the subqueries, and why I went about defining these models. It kept me from re-implementing the same behaviors in a way that might cause a regression. However, I am open to using the models already defined in `aggregation/models.py`, I'll just need to sit with the data a bit to come up with the right expectations. I think it might be best to keep the tests, like the subqueries, in place to make sure we maintain the coverage. If you think those behaviors are covered in other tests, though, then I'm open to cleaning them up.

charettes replied:
That makes sense, if they already exist we should definitely just port them over.

> However, I am open to using the models already defined in aggregation/models.py

If you could port them while adapting them to use the existing models that would be awesome.

charettes replied:
Also, if we're porting them over we should delete the original ones from `postgres_tests` given `postgres.aggregates.StringAgg` will only a shim at this point.

camuthig replied:
I did some work to merge the tests into the existing `AggregateTestCase`, keeping some of the more esoteric ones specific to the `StringAgg` behaviors but making them work with the existing data/models. I had to add a JSON field to the existing models to that end, but otherwise, I think it works pretty well.

charettes commented on tests/expressions_window/tests.py:
```diff
@@ -1706,14 +1706,14 @@ def test_fail_insert(self):
                 department="Management",
                 hire_date=datetime.date(2007, 7, 1),
-                salary=Window(expression=Sum(Value(10000), order_by=F("pk").asc())),
+                salary=Window(expression=Sum(Value(10000))),
```
Can you remind me why these changes are necessary?

camuthig replied:
I need to remind myself as well. I don't exactly remember writing this line. I'll refresh my memory and get back to you.

camuthig replied:
@charettes Oh, when I started validating `allow_order_by` in `Aggregate.__init__`, this test started throwing an error because `Sum` does not actually support order by. So I chose to remove the order by.

charettes replied:
That makes sense!

charettes commented on tests/aggregation/tests.py:
```diff
@@ -2359,3 +2390,256 @@ def test_aggregate_combined_queries(self):
+    def test_string_agg_distinct(self):
+        StringAggTestModel.objects.all().delete()
+
+        StringAggTestModel.objects.create(char_field="Foo")
+        StringAggTestModel.objects.create(char_field="Foo")
+        StringAggTestModel.objects.create(char_field="Bar")
```
Could use `bulk_create` here.

charettes commented on tests/aggregation/tests.py:
```diff
@@ -2359,3 +2390,256 @@ def test_aggregate_combined_queries(self):
+            (False, 2, 1),
+        ]:
+            with self.subTest(aggregate=f"distinct_{distinct_value}"):
+                # WIP I should brew install sqlite3 from my camuthig account in order to get
+                #   a more recent version that supports ordering group_concat
+                # WIP I'm getting the below, but only one argument is provided...
+                # SQLite doesn't support DISTINCT on aggregate functions accepting multiple arguments.
```
Left some pointers about how to best approach this problem.

charettes commented on tests/aggregation/tests.py:
```diff
@@ -2359,3 +2390,256 @@ def test_aggregate_combined_queries(self):
+        )
+        self.assertEqual(list(values), expected_result)
+
+    def test_filter_in_subquery_with_exclude(self):
```
I appreciate the testing but does this cover any particular edge case in subqueries?

charettes commented on tests/aggregation/tests.py:
```diff
@@ -2359,3 +2390,256 @@ def test_aggregate_combined_queries(self):
+    def test_default_argument(self):
+        StringAggTestModel.objects.all().delete()
+        tests = [
+            (StringAgg("char_field", delimiter=";", default="<empty>"), "<empty>"),
```
I would expect this to crash as `default` must be passed as `Value`? Passing `str` to `Func` should normally attempt resolve to field references (e.g. `"<empty>" -> F("<empty>")") and crash if missing.

camuthig replied:
> delimiter should not be wrapped in a Value as that prevent references to columns and normally string arguments are considered to be `F`'ield references.

The `Value` wrapper on a string comes right from the old implementation in the Postgres-specific version. Would it make more sense to allow for the string to come through and wrap it as a value for 5.2 with a deprecation warning in 6? That does break the ability to use a string as an alias for `F` for 5.2, though.

camuthig replied:
I think I have something together that wraps the `__init__` from the Postgres version of the `StringAgg` class and does the `Value(str(delimiter))` and logs a deprecation warning. The main `StringAgg` can then expect an `Expression` of some sort.

charettes replied:
> I think I have something together that wraps the __init__ from the Postgres version of the StringAgg class and does the Value(str(delimiter)) and logs a deprecation warning. The main StringAgg can then expect an Expression of some sort.

That's perfect, using the deprecation period for `postgres.StringAgg` to `models.StringAgg` seems like a great way to normalize this behavior 🎉

charettes reviewed (commented):
Thanks for the patch @camuthig!

I left a few comments and pointers. My main points of feedbacks are

1. `delimiter` should not be wrapped in a `Value` as that prevent references to columns and normally string arguments are considered to be `F`'ield references.
2. I appreciate the attention to testing but we should focus them on `StringAgg` behaviour and `order_by` support. Did you notice anything special when subqueries were involved?

charettes commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -49,17 +51,15 @@ class BoolOr(Aggregate):
+            "django.db.models.aggregate.StringAgg instead.",
+            category=RemovedInDjango60Warning,
+        )
+        super().__init__(expression, delimiter, **extra)
```
@camuthig to make sure I understand, this is where you would add the logic to pass `Value(delimiter_expr)` to `super()` with a deprecation warning?

camuthig replied:
@charettes I just pushed a change with this logic. Yes, this is where I was planning to do it.

camuthig commented:
@charettes I think I have updates for all of your comments at this phase. I was able to build a custom python environment with pyenv and brew using guidance [here](https://github.com/pyenv/pyenv/issues/333#issuecomment-234707676) and test the group concat and ordering behaviors with a more recent version of sqlite, so that is cool.

I know there are some conflicts and I can work on cleaning those up. I'm also not totally sure what is wrong with the docs. I ran the same command on my own machine, and everything seems to build properly.

charettes commented on tests/aggregation/tests.py:
```diff
@@ -2158,6 +2199,155 @@ def test_having_with_no_group_by(self):
+        }
+        self.assertEqual(values, expected_values)
+
+    @skipUnlessDBFeature("supports_aggregate_order_by_clause")
```
```suggestion
    @skipUnlessDBFeature("supports_json_field", "supports_aggregate_order_by_clause")
```

charettes commented on tests/aggregation/tests.py:
```diff
@@ -2158,6 +2199,155 @@ def test_having_with_no_group_by(self):
+        self.assertEqual(values, {"stringagg": "en,fr"})
+
+    def test_string_agg_filter_in_subquery(self):
+        """
+        This test is based on tests taken from existing PostgreSQL specific tests and
+        kept to avoid regressions as StringAgg is ported to the shared database module.
+        """
+
```
Good to point out in the PR review tab but they don't need to live in the tests themselves.

charettes commented on tests/aggregation/tests.py:
```diff
@@ -2158,6 +2199,155 @@ def test_having_with_no_group_by(self):
+                agg=Subquery(subquery),
+            )
+            .exclude(agg__isnull=True)
+            # Different engines treat null STRING_AGG differently, so excluding it for
+            # consistency.
```
Could you elaborate on that?

I seems that SQLite, Postgres, MySQL, and Oracle all return `NULL` which should translate to `None`?

1. [SQLite](https://dbfiddle.uk/fzC4JzJm)
2. [MySQL](https://dbfiddle.uk/jf5-JeSq)
3. [Postgres](https://dbfiddle.uk/62y1WY4y)
4. [Oracle](https://dbfiddle.uk/e0gP_WKJ)

Did you possibly run into issues because of a lack of ordering on the returned value and the fact some backends default to `NULLS LAST` instead of `NULLS FIRST`?

If that's the case you should use `assertQuerySetEqual(..., ordered=False)` or specify an `order_by(OrderBy('agg', nulls_last=True))`.

I think it'd be good to include at least one tests that ensures that `None` is returned on an empty not-NULL set.

camuthig replied:
You are right, I was reading the test results wrong. I went with `assertQuerySetEqual(..., ordered=False)` to resolve this.

charettes commented on tests/aggregation/tests.py:
```diff
@@ -2158,6 +2199,155 @@ def test_having_with_no_group_by(self):
+
+    @skipUnlessDBFeature("supports_aggregate_order_by_clause")
+    def test_order_by_in_subquery(self):
+        """
+        This test is based on tests taken from existing PostgreSQL specific tests and
+        kept to avoid regressions as StringAgg is ported to the shared database module.
+        """
```
```suggestion
```

charettes commented on tests/postgres_tests/test_aggregates.py:
```diff
@@ -151,12 +145,21 @@ def test_default_argument(self):
                     )
                     self.assertEqual(values, {"aggregation": expected_result})
 
+    def test_ordering_warns_of_deprecation(self):
+        msg = "The ordering argument is deprecated. Use order_by instead."
+        with self.assertWarnsMessage(RemovedInDjango60Warning, msg):
+            values = AggregateTestModel.objects.aggregate(
+                arrayagg=ArrayAgg("integer_field", ordering=F("integer_field").desc())
+            )
+            self.assertEqual(values, {"arrayagg": [2, 1, 0, 0]})
```
We'll need also need tests for the creation of `postgres.StringAgg`.

charettes replied:
The following should ensure that the proper `stacklevel` is passed.

```suggestion
        with self.assertWarnsMessage(RemovedInDjango60Warning, msg) ctx:
            values = AggregateTestModel.objects.aggregate(
                arrayagg=ArrayAgg("integer_field", ordering=F("integer_field").desc())
            )
            self.assertEqual(values, {"arrayagg": [2, 1, 0, 0]})
        self.assertEqual(ctx.filename, __file__)
```

charettes commented on django/contrib/postgres/aggregates/mixins.py:
```diff
@@ -1,9 +1,19 @@
 class OrderableAggMixin:
-    def __init__(self, *expressions, ordering=(), **extra):
+    def __init__(self, *expressions, ordering=(), order_by=(), **extra):
+        if ordering:
```
I think we should raise a `TypeError` if both are specified.

charettes reviewed (commented):
This is looking great, I left a few comments regarding null handling and a few other things.

I like that you started to breakdown the commits I may suggest a different order and naming scheme.

> Refs #35444 -- Deprecated contrib.postgres aggregates ordering for order_by.
>
> (Explain the reasoning to better align with Window.order_by and plans to move support for Aggregate)

> Fixed #35444 -- Add generic support for Aggregate.order_by.
>
> (Explain that this now supersedes the contrib.postgres support and introduces StringAgg to exercise this support)

^^ this commit should make `OrderableAggMixin` unused.

> Refs #35444 -- Deprecated contrib.postgres.OrderableAggMixin
>
> (Explain that it serves no purposes now that Aggregate supports order_by but it is kept around for a deprecation cycle as some users might be relying on it)

camuthig committed:
Refs #35444 -- Deprecated contrib.postgres.OrderableAggMixin.

This commit does not create any functional changes, but marks the
existing `OrderableAggMixin` class as deprecated so that developers
using it directly can be made aware of its future removal.

camuthig commented:
@charettes I have update the code to hit all of your comments and get the linters passing. I will take a dive into the failing tests shortly. At least some of them appear to be because we have moved the `connection.features.supports_aggregate_filter_clause` into the `AggregateFilter.as_sql` function, but have historically allowed for this check to fail and convert it to a `WHEN` statement. Now, though, when converting to the `WHEN`, we are throwing the `NotSupportedError` again. If you have thoughts on how you would like to see that flow, I'm open to suggestions.

charettes commented:
> At least some of them appear to be because we have moved the connection.features.supports_aggregate_filter_clause into the AggregateFilter.as_sql function, but have historically allowed for this check to fail and convert it to a WHEN statement. Now, though, when converting to the WHEN, we are throwing the NotSupportedError again. If you have thoughts on how you would like to see that flow, I'm open to suggestions.

The following should do

```diff
diff --git a/django/db/models/aggregates.py b/django/db/models/aggregates.py
index 6cf0bd9a60..d070944039 100644
--- a/django/db/models/aggregates.py
+++ b/django/db/models/aggregates.py
@@ -29,6 +29,10 @@ class AggregateFilter(Func):
     arity = 1
     template = " FILTER (WHERE %(expressions)s)"

+    @property
+    def condition(self):
+        return self.source_expressions[0]
+
     def as_sql(self, compiler, connection, **extra_context):
         if not connection.features.supports_aggregate_filter_clause:
             raise NotSupportedError
@@ -187,7 +191,7 @@ def as_sql(self, compiler, connection, **extra_context):
                 copy = self.copy()
                 copy.filter = None
                 source_expressions = copy.get_source_expressions()
-                condition = When(self.filter, then=source_expressions[0])
+                condition = When(self.filter.condition, then=source_expressions[0])
                 copy.set_source_expressions([Case(condition)] + source_expressions[1:])
                 return copy.as_sql(compiler, connection, **extra_context)
```

The problem was that we were building the fallback `When` with the `AggregateFilter` instead of its underlying `condition: Q`.

camuthig commented:
@charettes thanks for the recommendation. It works great. 

I spent time digging into the other failures in MySQL and determined that the MySQL implementation of `GROUP_CONCAT` doesn't actually play nicely with any of the other engine implementations. I have some code together that passed the aggregate tests on both MySQL and Postgres locally, but it is certainly messy and probably brittle.

First, it allows for multiple expressions, which we can ignore for now. The bigger issues are that MySQL doesn't support filtering on aggregates and it uses a different format order between the delimiter and ordering.

Sqlite: `GROUP_CONCAT(expression, delimiter ORDER BY order_by)`
MySQL: `GROUP_CONCAT(expressions* ORDER BY order_by SEPARATOR delimiter)`

So here, if you want to introduce a delimiter you MUST have a different template, which is solvable. The bigger issue is that the order matters between `ORDER BY` and `SEPARATOR`: `SEPARATOR delimiter ORDER BY order_by` is invalid syntax. This required me to manually build the parameters in `as_sql` and force the order of things to get the parameters into the right order depending on the engine.

This was further compounded by the fact MySQL doesn't support aggregate filtering and we are forcing a `CASE` statement. So if you have `filter`, `order_by`, and `delimiter` you need a third ordering of parameters, with the filter parameters coming before even the expressions. 😞 

Do you think there are some level of feature flags we can throw on MySQL to make this easier? Or maybe allow for engine-specific default separators to be ignored in our queries? I have a commit with some experimentation here: https://github.com/django/django/pull/18361/commits/34c8fe08cd15c62a5e74f1308babe01e0bd9dac0

charettes commented:
> I have a commit with some experimentation here: https://github.com/django/django/commit/cbe50126aa80ca8b5d79c48578aca1d29a673288

Thanks for pushing a commit demonstrating the scope of the problem.

Looking at it my immediate reaction would be to avoid over complicating `StringAgg.as_sql` and favor encapsulating the logic entirely in `as_mysql` instead. In there you'll know that the `FILTER` clause is not usable and will likely be able to simplify things quite a bit from some local testing locally.

> Or maybe allow for engine-specific default separators to be ignored in our queries?

I'm not sure I'm understanding what you mean here.

Now that the MySQL tests are passing I'll give Oracle a test run as well.

charettes commented:
buildbot, test on oracle.

camuthig commented:
@charettes do you have recommendations on how to test the Oracle engine myself? I am having trouble with the Oracle Docker images. Maybe cause I am on a M1 Mac? I'm not sure. It looks like a pretty common errors, so it seems some change I have made at a base level is causing the compiled queries to be incorrect but I can't really tell why without running against the database.

charettes commented:
@camuthig I usually try two approaches on my M1

1. I had success once I ran `brew install colima` and tried again
2. Low-tech approach is to attempt to generate the raw SQL locally and rely on tools such as [dbfiddle](https://dbfiddle.uk/Wh4zpMKJ).

I really wish CI ran with `--debug-sql` at least so we could peak at the generated SQL

Any suggestions @felixxm?

charettes commented:
I managed to get tests running on Oracle here's the non-localized error message

```
======================================================================
ERROR: test_annotate_values_list (aggregation.tests.AggregateTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/django/source/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
  File "/django/source/django/db/backends/oracle/base.py", line 577, in execute
    return self.cursor.execute(query, self._param_generator(params))
  File "/usr/local/lib/python3.10/site-packages/oracledb/cursor.py", line 710, in execute
    impl.execute(self)
  File "src/oracledb/impl/thin/cursor.pyx", line 196, in oracledb.thin_impl.ThinCursorImpl.execute
  File "src/oracledb/impl/thin/protocol.pyx", line 440, in oracledb.thin_impl.Protocol._process_single_message
  File "src/oracledb/impl/thin/protocol.pyx", line 441, in oracledb.thin_impl.Protocol._process_single_message
  File "src/oracledb/impl/thin/protocol.pyx", line 433, in oracledb.thin_impl.Protocol._process_message
  File "src/oracledb/impl/thin/messages.pyx", line 74, in oracledb.thin_impl.Message._check_and_raise_exception
oracledb.exceptions.DatabaseError: ORA-22848: cannot use NCLOB type as comparison key
Help: https://docs.oracle.com/error-help/db/ora-22848/

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/django/source/tests/aggregation/tests.py", line 939, in test_annotate_values_list
    self.assertEqual(list(books), [(self.b1.id, "159059725", 34.5)])
  File "/django/source/django/db/models/query.py", line 381, in __iter__
    self._fetch_all()
  File "/django/source/django/db/models/query.py", line 1909, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/django/source/django/db/models/query.py", line 229, in __iter__
    return compiler.results_iter(
  File "/django/source/django/db/models/sql/compiler.py", line 1536, in results_iter
    results = self.execute_sql(
  File "/django/source/django/db/models/sql/compiler.py", line 1585, in execute_sql
    cursor.execute(sql, params)
  File "/django/source/django/db/backends/utils.py", line 122, in execute
    return super().execute(sql, params)
  File "/django/source/django/db/backends/utils.py", line 79, in execute
    return self._execute_with_wrappers(
  File "/django/source/django/db/backends/utils.py", line 92, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/django/source/django/db/backends/utils.py", line 100, in _execute
    with self.db.wrap_database_errors:
  File "/django/source/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/django/source/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
  File "/django/source/django/db/backends/oracle/base.py", line 577, in execute
    return self.cursor.execute(query, self._param_generator(params))
  File "/usr/local/lib/python3.10/site-packages/oracledb/cursor.py", line 710, in execute
    impl.execute(self)
  File "src/oracledb/impl/thin/cursor.pyx", line 196, in oracledb.thin_impl.ThinCursorImpl.execute
  File "src/oracledb/impl/thin/protocol.pyx", line 440, in oracledb.thin_impl.Protocol._process_single_message
  File "src/oracledb/impl/thin/protocol.pyx", line 441, in oracledb.thin_impl.Protocol._process_single_message
  File "src/oracledb/impl/thin/protocol.pyx", line 433, in oracledb.thin_impl.Protocol._process_message
  File "src/oracledb/impl/thin/messages.pyx", line 74, in oracledb.thin_impl.Message._check_and_raise_exception
django.db.utils.DatabaseError: ORA-22848: cannot use NCLOB type as comparison key
Help: https://docs.oracle.com/error-help/db/ora-22848/
```

charettes commented on tests/aggregation/models.py:
```diff
@@ -30,6 +30,7 @@ class Book(models.Model):
     contact = models.ForeignKey(Author, models.CASCADE, related_name="book_contact_set")
     publisher = models.ForeignKey(Publisher, models.CASCADE)
     pubdate = models.DateField()
+    print_info = models.JSONField(null=True, default=None)
```
@camuthig this is the source of all your Oracle test failures...

Django uses a `NCLOB` column to persists `JSONField` and Oracle doesn't allow to `GROUP BY` a `NCLOB` field.

I don't think the usage of a `JSONField` is mandatory to test this feature so I'd suggest using a different field type instead.

charettes commented on tests/aggregation/tests.py:
```diff
@@ -2178,6 +2215,147 @@ def test_having_with_no_group_by(self):
+        }
+        self.assertEqual(values, expected_values)
+
+    @skipUnlessDBFeature("supports_json_field", "supports_aggregate_order_by_clause")
+    def test_string_agg_jsonfield_order_by(self):
+        values = Book.objects.aggregate(
+            stringagg=StringAgg(
+                KeyTextTransform("lang", "print_info"),
+                delimiter=Value(","),
+                order_by=KeyTextTransform("lang", "print_info"),
```
I guess this was ported from the Postgres tests.

I suggest creating a distinct model for this particular case instead.

charettes reviewed (commented):
Oracle failures were due to your addition of the `print_info` field.

felixxm commented:
> @camuthig I usually try two approaches on my M1
> 
>     1. I had success once I ran `brew install colima` and tried again
> 
>     2. Low-tech approach is to attempt to generate the raw SQL locally and rely on tools such as [dbfiddle](https://dbfiddle.uk/Wh4zpMKJ).
> 
> 
> I really wish CI ran with `--debug-sql` at least so we could peak at the generated SQL
> 
> Any suggestions @felixxm?

The "easiest" way is to use [VM provided by Oracle](https://www.oracle.com/database/technologies/databaseappdev-vm.html).

camuthig commented:
@charettes I have updated the code with two things:

1. Swapped the JSON test behaviors to a new, JSON specific model to avoid adverse effects on other tests.
2. Changed up how I dealt with MySQL to make it a bit more simple, and generally focused on just MySQL.

I haven't had a chance to set up the Oracle VM, so I don't know that the tests are going to pass on that backend. Can we run it again, and if issues continue, then I will get that VM configured.

I kept the MySQL changes as a separate commit to make them easier to review at this phase: https://github.com/django/django/pull/18361/commits/43415289e723fbcb7fc772f4c17718f7b72d30e1 . Let me know what you think of that. If we are happy with it, I will squash it back down.

felixxm commented:
buildbot, test on oracle.

charettes commented:
@camuthig that's awesome, thanks for the summary!

Sure we can give the Oracle suite a run and see how it goes. I have the VM setup locally now so I should be able to provide guidance as well if we run into issues.

charettes commented:
buildbot, test on oracle.

camuthig commented:
@charettes I still can't get oracle running locally. Not sure what is wrong with my system but the VM keeps crashing on me. I do believe I saw an error in the template for the Oracle SQL that was causing the failures on the order by queries and fixed it. Question for you: if I add the buildbot message myself, will it kick off the Oracle tests? And is it okay if I do that as I iterate? I'm not sure how that buildbot works and what is okay versus an over step.

charettes commented:
@camuthig I believe it should yes or that we can add you to an allow list otherwise. Could you trying `buildbot, test on oracle.`

camuthig commented:
buildbot, test on oracle.

camuthig commented:
Looks like the buildbot responded @charettes ! Thanks.

charettes commented:
One reported failure with the associated query

```
======================================================================
FAIL: test_string_agg_filter_in_subquery (aggregation.tests.AggregateTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/django/source/tests/aggregation/tests.py", line 2329, in test_string_agg_filter_in_subquery
    self.assertQuerySetEqual(expected_values, values, ordered=False)
  File "/django/source/django/test/testcases.py", line 1250, in assertQuerySetEqual
    return self.assertDictEqual(Counter(items), Counter(values), msg=msg)
AssertionError: Count[40 chars] 1, 'Paul Bissex;Wesley J. Chun': 1, 'Peter No[48 chars]: 1}) != Count[40 chars] 1, '': 1, 'Paul Bissex;Wesley J. Chun': 1, 'P[46 chars]: 1})
  Counter({'Adrian Holovaty': 1,
           'Brad Dayley': 1,
+          '': 1,
           'Paul Bissex;Wesley J. Chun': 1,
           'Peter Norvig;Stuart Russell': 1,
-          'Peter Norvig': 1,
?                           ^

+          'Peter Norvig': 1})
?                           ^^

-          None: 1})

----------------------------------------------------------------------
(0.021)
SELECT
    (
        SELECT
            LISTAGG (
                CASE WHEN NOT (U2."NAME" LIKE TRANSLATE(J % USING NCHAR_CS) ESCAPE TRANSLATE('\' USING NCHAR_CS)
                    AND U2."NAME" IS NOT NULL) THEN
                            U2."NAME"
                ELSE
                    NULL
                        END,;

) AS "AGG"
FROM
    "AGGREGATION_BOOK" U0
    LEFT OUTER JOIN "AGGREGATION_BOOK_AUTHORS" U1 ON (U0."ID" = U1."BOOK_ID")
    LEFT OUTER JOIN "AGGREGATION_AUTHOR" U2 ON (U1."AUTHOR_ID" = U2."ID")
WHERE
    U0."ID" = ("AGGREGATION_BOOK"."ID")
GROUP BY
    U0."ID",
    U0."ISBN",
    U0."NAME",
    U0."PAGES",
    U0."RATING",
    U0."PRICE",
    U0."CONTACT_ID",
    U0."PUBLISHER_ID",
    U0."PUBDATE") AS "AGG"
FROM
    "AGGREGATION_BOOK";

 args=('J%', ';'); ALIAS=DEFAULT
```

charettes commented:
I think we'll have no choice but not to support `StringAgg(filter)` on Oracle (we could raise `NotSupportedError` in `as_oracle` given `self.filter`) because `LISTAGG` doesn't offer a way to filter out values and Oracle doesn't support the `FILTER (WHERE ...)` clause which forces the usage of `CASE WHEN` which yields `NULL` on mismatches.

As another PR ran into (https://github.com/django/django/pull/18351#pullrequestreview-2287172954) in order to support this properly aggregates would need to support being turned into subquery aggregates (https://github.com/django/django/pull/11841) which is way beyond the scope of this PR.

I'm not sure what's causing the GIS failures but the error message is

```
django.db.utils.DatabaseError: ORA-00907: missing right parenthesis
Help: https://docs.oracle.com/error-help/db/ora-00907/
```

And the generated SQL for `test_geoagg_subquery` is 

```
SELECT 
  SDO_UTIL.TO_WKBGEOMETRY(
    SDO_AGGR_UNION() SDOAGGRTYPE("__COL1", NULL, % s)
  ) 
FROM 
  (
    SELECT 
      DISTINCT "GEOAPP_CITY"."ID" AS "COL1", 
      "GEOAPP_CITY"."NAME" AS "COL2", 
      "GEOAPP_CITY"."POINT" AS "COL3", 
      "GEOAPP_CITY"."POINT" AS "__COL1" 
    FROM 
      "GEOAPP_CITY" 
    WHERE 
      SDO_INSIDE(
        "GEOAPP_CITY"."POINT", 
        SDO_GEOMETRY(% s, 4326)
      ) = 'TRUE'
  ) subquery
```

on your branch while it's

```sql
SELECT 
  SDO_UTIL.TO_WKBGEOMETRY(
    SDO_AGGR_UNION() SDOAGGRTYPE("__COL1", NULL, % s)
  ) 
FROM 
  (
    SELECT 
      DISTINCT "GEOAPP_CITY"."ID" AS "COL1", 
      "GEOAPP_CITY"."NAME" AS "COL2", 
      "GEOAPP_CITY"."POINT" AS "COL3", 
      "GEOAPP_CITY"."POINT" AS "__COL1" 
    FROM 
      "GEOAPP_CITY" 
    WHERE 
      SDO_INSIDE(
        "GEOAPP_CITY"."POINT", 
        SDO_GEOMETRY(MULTIPOLYGON(%s), 4326)
      ) = 'TRUE'
  ) subquery
```

on main

I highly suspect `GeoAggregate.as_oracle` needs adjustments with regards to `self.filter`.

charettes commented:
@camuthig I think the following should address the Oracle GIS failures

```diff
diff --git a/django/contrib/gis/db/models/aggregates.py b/django/contrib/gis/db/models/aggregates.py
index abff1a5206..e526a8d3ef 100644
--- a/django/contrib/gis/db/models/aggregates.py
+++ b/django/contrib/gis/db/models/aggregates.py
@@ -33,15 +33,14 @@ def as_oracle(self, compiler, connection, **extra_context):
         if not self.is_extent:
             tolerance = self.extra.get("tolerance") or getattr(self, "tolerance", 0.05)
             clone = self.copy()
-            source_expressions = self.get_source_expressions()
-            source_expressions.pop()  # Don't wrap filters with SDOAGGRTYPE().
+            *source_expressions, filter_expr, order_by_expr = self.get_source_expressions()
             spatial_type_expr = Func(
                 *source_expressions,
                 Value(tolerance),
                 function="SDOAGGRTYPE",
                 output_field=self.output_field,
             )
-            source_expressions = [spatial_type_expr, self.filter]
+            source_expressions = [spatial_type_expr, filter_expr, order_by_expr]
             clone.set_source_expressions(source_expressions)
             return clone.as_sql(compiler, connection, **extra_context)
         return self.as_sql(compiler, connection, **extra_context)
```

camuthig commented:
@charettes I pulled in your change and will run it against the Oracle here. I really appreciate your help on that one, I wasn't going to be able to triage that one quickly.

Regarding the `test_string_agg_filter_in_subquery` the challenge here is that Oracle is treating these case clauses differently than, for example, MySQL. MySQL is treating those values as `None` and we are getting the same results for MySQL as PostgreSQL. Oracle appears to be casting `null` to a string and getting `''` instead. I'm good with blocking the filter behavior for any backends that don't support filter natively. I think that check would be pretty straightforward to add in there and would cover MySQL and Oracle.

charettes commented:
> Oracle appears to be casting null to a string and getting '' instead. 

@camuthig the reason for this is this is that [for Oracle](https://github.com/django/django/blob/f4813211e2d8017b56b7447f56ad17df3fae9aa3/django/db/backends/oracle/features.py#L16) `NULL` and `""` (empty string) are the same thing which means that it doesn't provide facilities to differentiate between the two (it can't).

Thinking more about it I think that in order to account for that we should make the following adjustments

```diff
diff --git a/tests/aggregation/tests.py b/tests/aggregation/tests.py
index a23ae47487..fd535b1867 100644
--- a/tests/aggregation/tests.py
+++ b/tests/aggregation/tests.py
@@ -2323,7 +2323,7 @@ def test_string_agg_filter_in_subquery(self):
             "Paul Bissex;Wesley J. Chun",
             "Peter Norvig;Stuart Russell",
             "Peter Norvig",
-            None,
+            "" if connection.features.interprets_empty_strings_as_nulls else None,
         ]

         self.assertQuerySetEqual(expected_values, values, ordered=False)
```

Instead of completely disabling filtering or Oracle. The values are filtered, Oracle just doesn't offer a way to differentiate the kind of _emptyness_ its dealing with.

camuthig commented:
@charettes that is really interesting to know about Oracle. I like that strategy, so I've put it in and will get it all run. Hopefully we'll go green on this round!

charettes commented:
buildbot, test on oracle.

camuthig commented:
@charettes any thoughts on the one server side binding postgres error? It doesn't reproduce for me locally, and I'm not sure how the changes in this branch could have started impacting update/create behaviors in transactions at this phase. Is there chance it was a flakey failure or something?

charettes commented on django/db/models/aggregates.py:
```diff
@@ -196,6 +256,93 @@ def _get_repr_options(self):
+        else:
+            template = "%(function)s(%(distinct)s%(expressions)s)%(filter)s"
+
+        return super().as_sql(
```
I think that both `super().as_sql` and `self.as_sql` resolve to the same since `StringAgg` doesn't override `as_sql`?

```suggestion
        return self.as_sql(
```

charettes commented on django/db/models/aggregates.py:
```diff
@@ -196,6 +256,93 @@ def _get_repr_options(self):
+                **extra_context,
+            )
+
+        return super().as_sql(compiler, connection, **extra_context)
```
```suggestion
        return self.as_sql(compiler, connection, **extra_context)
```

charettes reviewed (commented):
@camuthig I think it can be safely ignored yes it seems a flaky failure due to CI setup which means we've got all the tests passing 🎉 

So at this point the only parts missing are

1. Fold commits 4,5.
2. Document the addition of `Aggregate(order_by)` and `allow_order_by` [here](https://github.com/django/django/blob/16af0c60ed90813ed70ecef4e64a89301ed0f0a2/docs/ref/models/expressions.txt#L453) and add a release note about it [here](https://github.com/django/django/blob/16af0c60ed90813ed70ecef4e64a89301ed0f0a2/docs/releases/5.2.txt#L208-L224)
3. Document `StringAgg` [here](https://github.com/django/django/blob/16af0c60ed90813ed70ecef4e64a89301ed0f0a2/docs/ref/models/querysets.txt#L4047) and add a 5.2 release note about it as well.

camuthig commented:
buildbot, test on oracle.

camuthig commented:
@charettes is there anything I need to do to pull this forward in the review process? I know it was DjangoCon last week, so people were probably super busy. I just want to make sure I'm not forgetting part of the process.

camuthig commented:
Thanks @nessita ! I think I have everything configured properly to get it into the queue now.

charettes commented on django/db/models/aggregates.py:
```diff
@@ -115,42 +166,51 @@ def get_group_by_cols(self):
+            and not connection.features.supports_aggregate_distinct_multiple_argument
+            and len(super().get_source_expressions()) > 1
+        ):
+            raise NotSupportedError
```
Minor but I think we'll want to raise a more explicit message here as it might be user facing


```suggestion
            raise NotSupportedError(
                f"{self.name} doesn't support `distinct` with multiple expressions on {connection.vendor}"
            )
```

We should also add a `@skipIfDBFeature("supports_aggregate_distinct_multiple_argument")` test for it that `assertRaisesMessage`.

charettes reviewed (commented):
The patch is in a pretty good state @camuthig thanks for your efforts here. I left some comments about testing the lack of support for multiple expression distinct on SQLite but otherwise it seems ready for an eventual final review.

You could also change the PR title to match the commit message format (e.g. `Fixed #35444 -- Added generic support for Aggregate.order_by`) but that's very minor.

camuthig commented:
@nessita I wanted to follow up on this pull request. I was traveling for a few weeks, so it fell of the radar to check in on it. I believe the requested changes are all there, and I can clean up any doc conflicts as needed.

I don't know the 5.2 release schedule, but do you believe we can get this in there or is it going to be better to wait for 6.0?

charettes commented on django/contrib/gis/db/models/aggregates.py:
```diff
@@ -33,15 +33,14 @@ def as_oracle(self, compiler, connection, **extra_context):
-            source_expressions = self.get_source_expressions()
-            source_expressions.pop()  # Don't wrap filters with SDOAGGRTYPE().
+            *source_expr, filter_expr, order_by_expr = self.get_source_expressions()
             spatial_type_expr = Func(
-                *source_expressions,
+                *source_expr,
```
Since it might contain more than one it'd switch to `source_exprs`.

charettes reviewed (approved):
> Am I reading you latest comment correctly in that this has an implicit/soft approval from your end?

@nessita that's a fair assessment I think this path is ready for a final review!

camuthig commented:
@charettes hopefully just one last commit based on failing tests from the composite PK changes (which I am _really_ excited about) that were brought in once I rebased from main.

The new `ColPairs` type used in aggregates, like `Count`, was essentially taking the filter that was originally passed into the aggregate class, which was then wrapped in `AggregateFilter`, and wrapping it again. So we have like `AggregateFilter(AggregateFilter(stuff))`. This results in the template generating twice and creating bad SQL.

https://github.com/django/django/pull/18361/commits/9027c537c6523c388a8ad9aa32b881c04d6760e5

I made some changes in the above commit to allow for receiving an instance of `AggregateFilter` and avoid the extra wrapping. I'm open for your thoughts on how this change is written to make them match the project. I'm hoping this will have the tests passing and once we get it write, I'll rebase it into the clean commits.

charettes commented:
@camuthig thanks for rebasing and looking into the composite primary key support conflicts, we're all excited about this feature landing!

I think that a more adequate solution would be

```diff
diff --git a/django/db/models/aggregates.py b/django/db/models/aggregates.py
index a9fbd86753..0efc17d57b 100644
--- a/django/db/models/aggregates.py
+++ b/django/db/models/aggregates.py
@@ -37,13 +37,6 @@ class AggregateFilter(Func):
     arity = 1
     template = " FILTER (WHERE %(expressions)s)"

-    @classmethod
-    def from_param(cls, param):
-        if isinstance(param, AggregateFilter):
-            return param
-        else:
-            return cls(param)
-
     def as_sql(self, compiler, connection, **extra_context):
         if not connection.features.supports_aggregate_filter_clause:
             raise NotSupportedError
@@ -96,7 +89,7 @@ def __init__(
             raise TypeError(f"{self.__class__.__name__} does not allow default.")

         self.distinct = distinct
-        self.filter = filter and AggregateFilter.from_param(filter)
+        self.filter = filter and AggregateFilter(filter)
         self.order_by = order_by
         self.default = default
         self.order_by = order_by and AggregateOrderBy.from_param(
@@ -254,17 +247,16 @@ def __init__(self, expression, filter=None, **extra):

     def resolve_expression(self, *args, **kwargs):
         result = super().resolve_expression(*args, **kwargs)
-        expr = result.source_expressions[0]
+        source_expressions = result.get_source_expressions()

         # In case of composite primary keys, count the first column.
-        if isinstance(expr, ColPairs):
+        if isinstance(expr := source_expressions[0], ColPairs):
             if self.distinct:
                 raise NotSupportedError(
                     "COUNT(DISTINCT) doesn't support composite primary keys"
                 )
-
-            cols = expr.get_cols()
-            return Count(cols[0], filter=result.filter)
+            source_expressions[0] = expr.get_cols()[0]
+            result.set_source_expressions(source_expressions)

         return result
```

which can be included in the `Fixed #35444 -- Added generic support for Aggregate.order_by.` commit

camuthig commented:
@charettes I appreciate the input. I've got it cleaned up and everything looks good to go now.

charettes commented:
Thanks for seeing this through @camuthig, the PR should be ready for a final review when Django fellows get to it. It might possible get in the 5.2 release window but if it's not the case it will surely make it in 6.0 given the advanced state of the work.

If that's something that means something to do don't hesitate to add yourself to the `AUTHORS` file, you're more than deserved your name in there!

sarahboyce commented on django/db/backends/base/features.py:
```diff
@@ -257,6 +257,15 @@ class BaseDatabaseFeatures:
     # expressions?
     supports_aggregate_filter_clause = False
 
+    # Does the database support ORDER BY in aggregate expressions?
+    supports_aggregate_order_by_clause = False
+
+    # Does the database backend support DISTINCT when using multiple arguments in an
+    # aggregate expression? For example, Sqlite treats the "delimiter" argument of
+    # STRING_AGG/GROUP_CONCAT as an extra argument and does not allow using a custom
+    # delimiter along with DISTINCT.
+    supports_aggregate_distinct_multiple_argument = True
```
I think these need release notes

sarahboyce replied:
I think I'm wrong, ignore this!

sarahboyce commented on docs/releases/5.2.txt:
```diff
@@ -466,3 +471,11 @@ Miscellaneous
 * The fallback to ``request.user`` when ``user`` is ``None`` in
   ``django.contrib.auth.login()`` and ``django.contrib.auth.alogin()`` will be
   removed.
```
The depredations need to be added to `docs/internals/deprecation.txt`

camuthig replied:
I have added some documentation around the items to be removed. Please let me know if it meets the needs/standards.

sarahboyce reviewed (commented):
Thank you for this work @camuthig :star: 

I've given an initial review mostly around the docs
Can you also resolve the conflicts?

camuthig committed:
Fixed #35444 -- Added generic support for Aggregate.order_by.

This moves the behaviors of `order_by` used in Postgres aggregates into
the `Aggregate` class. This allows for creating aggregate functions that
support this behavior across all database engines. This is shown by
moving the `StringAgg` class into the shared `aggregates` module and
adding support for all databases. The Postgres `StringAgg` class is now
a thin wrapper on the new shared `StringAgg` class.

Thank you Simon Charette for the review.

camuthig commented:
@sarahboyce I believe the requested tests are all added and the documentation should be good. The failing test doesn't mean anything to me. My guess is it is unrelated to this PR, but let me know if there is something else to work on.

sarahboyce reviewed (commented):
I have merged the first commit https://github.com/django/django/commit/d734f1651ccc0a74325f7b55f7eecc68edef6453
Resolved conflicts and forced pushed updates here

I need to do a review of the deprecation to check I understand what we're ending up with and hopefully the tests are still passing after the conflict resolution :crossed_fingers: 
Note that the feature freeze is Jan 15th, still optimistic for this one but there's a chance that this will have to be in 6.0

charettes commented on django/contrib/postgres/aggregates/mixins.py:
```diff
@@ -1,15 +1,10 @@
 import warnings
```
So something like?

```suggestion
# RemovedInDjango61Warning: When the deprecation ends, remove completely.
import warnings
```

sarahboyce replied:
Yes :+1: assuming I understand the plan correctly :grin:

charettes replied:
yeah I think that adds up, one the deprecation period ends there will be no need for specialized Postgres aggregate mixins.

sarahboyce commented on django/db/models/aggregates.py:
```diff
@@ -18,42 +27,92 @@
+
+    def as_sql(self, compiler, connection, **extra_context):
+        if not connection.features.supports_aggregate_filter_clause:
+            raise NotSupportedError(
+                "Aggregate filter clauses are not supported on this database backend."
+            )
```
I think this is missing a test

camuthig replied:
@sarahboyce we are actually catching this exception and using it in the `Aggregate.as_sql`, which is why there isn't a specific test case for it.

https://github.com/django/django/pull/18361/files#diff-c7d87b6e6b973bf87e5685442a6c1da9531c24bcaa86fe158c40ea0276986f32R204

charettes replied:
Right, we used this approach as it makes the management of source expressions for backends that don't support filtered aggregates easier to reason about.

sarahboyce replied:
Ah thank you!

sarahboyce commented on docs/ref/models/expressions.txt:
```diff
@@ -493,6 +502,11 @@ The ``filter`` argument takes a :class:`Q object <django.db.models.Q>` that's
 used to filter the rows that are aggregated. See :ref:`conditional-aggregation`
 and :ref:`filtering-on-annotations` for example usage.
 
+The ``order_by`` argument takes a field name (with an optional ``"-"`` prefix
+which indicates descending order) or an expression (or a tuple or list of
+strings and/or expressions) that specifies the ordering of the elements in the
+result.
```
I couldn't spot tests around a tuple/list of fields - can you check/add tests for these?

I also wonder if we can say that this is similar to what you can input to `order_by()`

charettes replied:
> I also wonder if we can say that this is similar to what you can input to order_by()

It should effectively accept similar inputs.

camuthig replied:
I have updated the documentation to include something about this. Let me know if you would like any changes.

sarahboyce commented:
buildbot, test on oracle.

sarahboyce commented on docs/releases/5.2.txt:
```diff
@@ -334,6 +334,15 @@ Models
+* The new :attr:`.Aggregate.allow_order_by` class attribute determines whether
+  the aggregate function allows passing an ``order_by`` keyword argument.
+
+* The new :class:`~django.db.models.StringAgg` aggregate returns the input
+  values concatenated into a string, separated by the ``delimiter`` string.
```
I wonder if we should mention that this used to only be supported on Postgres

camuthig replied:
I have updated the documentation to include something about this. Let me know if you would like any changes. (also moved to 6.0.txt)

sarahboyce commented on docs/releases/5.2.txt:
```diff
@@ -334,6 +334,15 @@ Models
+* The new :attr:`.Aggregate.allow_order_by` class attribute determines whether
+  the aggregate function allows passing an ``order_by`` keyword argument.
+
+* The new :class:`~django.db.models.StringAgg` aggregate returns the input
+  values concatenated into a string, separated by the ``delimiter`` string.
```
I wonder if we should mention that this aggregate used to only be supported on Postgres

sarahboyce reviewed (commented):
Pushed a couple of docs tweaks and some more deprecation comments

charettes commented:
@camuthig thanks for your continued efforts here. The changes unfortunately didn't make it in time for 5.2 feature freeze so you'll need to repoint at 6.0 on top of the already provided comments. Let me know if I can be of any help in pushing this forward.

camuthig commented on tests/postgres_tests/test_aggregates.py:
```diff
@@ -158,8 +154,9 @@ def test_ordering_warns_of_deprecation(self):
+    # RemovedInDjango61Warning: Remove this test
     def test_ordering_and_order_by_causes_error(self):
-        with self.assertWarns(RemovedInDjango61Warning):
+        with warnings.catch_warnings(record=True, action="always") as wm:
```
Things got a little weird here when I update the warnings to use 7.0. The issue being that interacting with the class this way triggers both a 6.1 warning (for the `order_by` usage) and 7.0 warnings (for using the PostgreSQL StringAgg). I couldn't get `assertWarns` to work with both getting thrown so dug in and tried using the underlying behaviors of it to get it down.

Let me know if there is a better way to write this, and I will refactor it.

charettes replied:
I think the way you did it is great, it's a peculiar case given it triggers two distinct warnings of different categories 👍

camuthig commented:
I think everything should be up to date and marked as deprecated as of 6.0 and to be removed in 7.0. I updated the documentation as requested as well.

I ran into testing issues around the layers of deprecation warnings. I have a working solution for it, but I'm not sure it is the right pattern.

sarahboyce commented:
buildbot, test on oracle.

charettes commented:
I should be able to have a look at it shortly. Looks at the test failures I assume this is happening because some aggregate is stashing `tuple` instances in `source_expressions` which shouldn't happen (only `Resolvable | None` should be allowed).

The `is None` is important as otherwise if a resolvable with a `__bool__` causing side-effect is passed it will be evaluated. The classic example is `QuerySet` which evaluates its query.

charettes commented on django/db/models/aggregates.py:
```diff
@@ -18,42 +27,92 @@
         self.distinct = distinct
-        self.filter = filter
+        self.filter = filter and AggregateFilter(filter)
+        self.order_by = order_by
         self.default = default
+        self.order_by = order_by and AggregateOrderBy.from_param(
+            f"{self.__class__.__name__}.order_by", order_by
+        )
```
Something when wrong with the rebase here, we assign to `self.order_by` twice?

charettes commented:
@sarahboyce I haven't figured out in which commit these should be folded but here are the required changes to get the tests passing

```diff
diff --git a/django/contrib/postgres/aggregates/mixins.py b/django/contrib/postgres/aggregates/mixins.py
index a6849c3930..8d3e40177b 100644
--- a/django/contrib/postgres/aggregates/mixins.py
+++ b/django/contrib/postgres/aggregates/mixins.py
@@ -16,7 +16,7 @@ def __init__(self, *expressions, ordering=(), order_by=(), **extra):
             if order_by:
                 raise TypeError("Cannot specify both order_by and ordering.")
             order_by = ordering
-
+        order_by = order_by or None
         super().__init__(*expressions, order_by=order_by, **extra)


diff --git a/django/db/models/aggregates.py b/django/db/models/aggregates.py
index 8b644e3599..48e00aa946 100644
--- a/django/db/models/aggregates.py
+++ b/django/db/models/aggregates.py
@@ -95,7 +95,6 @@ def __init__(

         self.distinct = distinct
         self.filter = filter and AggregateFilter(filter)
-        self.order_by = order_by
         self.default = default
         self.order_by = order_by and AggregateOrderBy.from_param(
             f"{self.__class__.__name__}.order_by", order_by
```

charettes commented:
We might want to also consider having `OrderByList.from_params` return `None` when passed an empty `tuple | list` to prevent similar crashes when `Aggregate(order_by)` is used directly instead.

```patch
diff --git a/django/db/models/aggregates.py b/django/db/models/aggregates.py
index 8b644e3599..a791593508 100644
--- a/django/db/models/aggregates.py
+++ b/django/db/models/aggregates.py
@@ -95,9 +95,8 @@ def __init__(

         self.distinct = distinct
         self.filter = filter and AggregateFilter(filter)
-        self.order_by = order_by
         self.default = default
-        self.order_by = order_by and AggregateOrderBy.from_param(
+        self.order_by = AggregateOrderBy.from_param(
             f"{self.__class__.__name__}.order_by", order_by
         )
         super().__init__(*expressions, **extra)
diff --git a/django/db/models/expressions.py b/django/db/models/expressions.py
index 68e5d2667e..ced5557732 100644
--- a/django/db/models/expressions.py
+++ b/django/db/models/expressions.py
@@ -1466,7 +1466,11 @@ def __init__(self, *expressions, **extra):

     @classmethod
     def from_param(cls, context, param):
+        if param is None:
+            return None
         if isinstance(param, (list, tuple)):
+            if not param:
+                return None
             return cls(*param)
         elif isinstance(param, str) or hasattr(param, "resolve_expression"):
             return cls(param)
@@ -1937,8 +1941,7 @@ def __init__(
                 self.partition_by = (self.partition_by,)
             self.partition_by = ExpressionList(*self.partition_by)

-        if self.order_by is not None:
-            self.order_by = OrderByList.from_param("Window.order_by", self.order_by)
+        self.order_by = OrderByList.from_param("Window.order_by", self.order_by)
         super().__init__(output_field=output_field)
         self.source_expression = self._parse_expressions(expression)[0]
```

sarahboyce commented:
buildbot, test on oracle.

sarahboyce commented:
buildbot, test on oracle.

sarahboyce reviewed (approved):
Thank you! Great job on this :star:

charettes commented:
Thank you for sticking around and seeing this through @camuthig, that was quite the 10 months adventure. Hopefully it was a valuable experience for you!

camuthig commented:
Woo! We did it! 

Thanks for all the help along the way, @charettes . It's definitely been a good experience for me. I learned a lot about how the ORM is working behind the scenes, which was my own goal. Now that we have this part done, I look forward to moving some of the other Postgres-specific functions over to general support. At least `ArrayAgg` but maybe something of the JSON agg too.

timgraham commented on tests/aggregation/tests.py:
```diff
@@ -2202,6 +2250,167 @@ def test_aggregate_arity(self):
+                )
+            )
+
+    def test_string_agg_filter(self):
+        values = Book.objects.aggregate(
+            stringagg=StringAgg(
+                "name",
+                delimiter=Value(";"),
+                filter=Q(name__startswith="P"),
+            )
+        )
+
+        expected_values = {
+            "stringagg": "Practical Django Projects;"
+            "Python Web Development with Django;Paradigms of Artificial "
+            "Intelligence Programming: Case Studies in Common Lisp",
+        }
+        self.assertEqual(values, expected_values)
```
As far as I see, `STRING_AGG()` doesn't define an ordering in the absence of an ORDER BY, so many of these assertions could be relaxed regarding the order in which the values are concatenated with ";". Did I miss something?

charettes replied:
@timgraham you're right, when ordering is not explicitly specified it is undefined so the test should be adjusted to compare the `;` split set of values.

timgraham replied:
Thanks. I've proposed https://github.com/django/django/pull/21536.
